crates/solvers/src/domain/solver/dex/fills.rs

```diff
@@ -0,0 +1,135 @@
+    /// amount to not eventually converge at 0.
+    smallest_fill: BigDecimal,
+}
+
+const ETH: eth::TokenAddress = eth::TokenAddress(eth::H160([0xee; 20]));
+
+impl Fills {
+    pub fn new(smallest_fill: eth::Ether) -> Self {
+        Self {
+            amounts: Default::default(),
+            smallest_fill: conv::ether_to_decimal(&smallest_fill),
+        }
+    }
+
+    /// Returns which dex query should be tried for the given order. Takes
+    /// information of previous partial fill attempts into account.
+    pub fn dex_order(
+        &self,
+        order: &order::Order,
+        prices: &dex::slippage::Prices,
+    ) -> Option<dex::Order> {
+        if !order.partially_fillable {
+            return Some(dex::Order::new(order));
+        }
+
+        let (token, total_amount) = match order.side {
+            order::Side::Buy => (order.buy.token, order.buy.amount),
+            order::Side::Sell => (order.sell.token, order.sell.amount),
+        };
+
+        let smallest_fill =
+            self.smallest_fill.clone() * prices.0.get(&token)? / prices.0.get(&ETH)?;
+        let smallest_fill = conv::decimal_to_ether(&smallest_fill)?.0;
+        tracing::trace!(?smallest_fill, ?order, "least amount worth filling");
+
+        let now = Instant::now();
+
+        let amount = match self.amounts.lock().unwrap().entry(order.uid) {
+            std::collections::hash_map::Entry::Vacant(entry) => {
+                entry.insert(CacheEntry {
+                    next_amount: total_amount,
+                    last_requested: now,
+                });
+                total_amount
+            }
+            std::collections::hash_map::Entry::Occupied(mut entry) => {
+                let entry = entry.get_mut();
+                entry.last_requested = now;
+
+                entry.next_amount = if entry.next_amount < smallest_fill {
+                    tracing::trace!("target fill got too small; starting over");
+                    total_amount
+                } else if entry.next_amount > total_amount {
+                    tracing::trace!("partially filled; adjusting to new total amount");
+                    total_amount
+                } else {
+                    // The regular case where we do our normal attempt.
+                    entry.next_amount
+                };
+
+                entry.next_amount
+            }
+        };
+
+        if amount < smallest_fill {
+            tracing::trace!(?amount, "order no longer worth filling");
+            return None;
+        }
+
+        let (sell, buy) = match order.side {
+            order::Side::Buy => (order.sell, eth::Asset { token, amount }),
+            order::Side::Sell => (eth::Asset { token, amount }, order.buy),
+        };
+
+        tracing::trace!(?amount, "trying to partially fill order");
+        Some(dex::Order::new(&order::Order {
+            sell,
+            buy,
+            ..*order
+        }))
```

Comment:
This syntax (specifying some fields and taking the remaining fields from a destructured object) is quite common in Rust and a some other languages so I would rather leave it as it is.

Reply:
It's not really the syntax that I find confusing, but the fact that you generate an intermediate `order::Order` that does not have an actual meaning but is just a small hack to set the parameters of the desired `dex::Order` more easily. The intermediate `order::Order` (say in the case of a sell order for simplicity) sells `partial_amount_sell_token` in exchange for `full_amount_buy_token`, which was weird to me and I went looking at how the called functions work to make sure it wasn't an issue. This doesn't matter in the logic because `dex::Order::new` drops any information on `full_amount_buy_token`.
Anyway, I wouldn't consider this issue something that necessarily needs addressing and I'm ok with keeping everything as it is.